File: java/2015/Day01.java
import module java.base;

public class Day01 {
    public static void main(String[] args) {
        var input = readInput("../../data/2015/day01.dat");
        if (input.isEmpty()) {
            return;
        }

        System.out.println("Part 1: " + solvePart1(input));
        System.out.println("Part 2: " + solvePart2(input));
    }

    private static String readInput(String filePath) {
        try {
            return Files.readString(Paths.get(filePath)).trim();
        } catch (NoSuchFileException e) {
            System.out.println("Data file missing: " + filePath);
        } catch (IOException ex) {
        }
        return "";
    }

    private static int solvePart1(String input) {
        return input.chars().map(c -> c == '(' ? 1 : -1).sum();
    }

    private static int solvePart2(String input) {
        var floor = 0;
        var position = 0;
        for (var c : input.toCharArray()) {
            position++;
            floor += (c == '(') ? 1 : -1;
            if (floor == -1) {
                return position;
            }
        }
        return floor;
    }
}

File: Csharp/2015/Day01/Day01.cs
using System;
using System.IO;

namespace Day01
{
    class Program
    {
        static void Main(string[] args)
        {
            string filePath = "../../data/2015/day01.dat";

            int floor = 0;
            int? position = null;
            int i = 0;

            using (StreamReader reader = new(filePath))
            {
                int c;
                while ((c = reader.Read()) != -1)
                {
                    char ch = (char)c;
                    if (ch == ')')
                    {
                        floor--;
                    }
                    else if (ch == '(')
                    {
                        floor++;
                    }

                    i++;

                    if (floor == -1 && position == null)
                    {
                        position = i;
                    }
                }
            }

            Console.WriteLine($"2015 Day 01 Part 1: {floor}");
            Console.WriteLine($"2015 Day 01 Part 2: {position}");
        }
    }
}
